### PR TITLE
Advanced changes in bmfuncts package

### PR DESCRIPTION
**Pull request by TickyWill**
Main changes deals with author affiliation parsing for identification of country and normalized institutions to which the author belongs. In addition, functions for impact factors attributions have been consolidated (ex: split into several functions of too complex functions).

### CODE REVIEW

**Comment by BaptisteRefalo** [comment]
Je viens juste de commencer à regarder mais j'aurais déjà une remarque: ça pourrait être intéressant de profiter de cette Pull Request pour compléter les docstrings vides dans le code 😊

**Comment by TickyWill** [comment]
Oui tu as raison. Ces docstrings etaient mon objectif premier mais il y a eu mon épaule et les corrections à faire…  Envoyé de mon iPhoneLe 30 août 2024 à 16:27, Baptiste Refalo ***@***.***> a écrit :﻿
Je viens juste de commencer à regarder mais j'aurais déjà une remarque: ça pourrait être intéressant de profiter de cette Pull Request pour compléter les docstrings vides dans le code 😊

File: bmfuncts/update_impact_factors.py [hunk 1]
```diff
@@ -233,18 +249,82 @@ def _formatting_wb_sheet(year, df, wb, first, if_database):
...
+                              most_recent_year_if_db_df,
+                              wb, first, if_database)
+    return wb 
+
+
+def update_inst_if_database(institute, org_tup, bibliometer_path, corpi_years_list):
+    """
+    Note:
+    
+    """
```
Comment (BaptisteRefalo) [inline bmfuncts/update_impact_factors.py:270]: TODO: complete the docstring :)

File: bmgui/parse_corpus_page.py [hunk 2]
```diff
@@ -292,7 +294,11 @@ def _launch_parsing(self, master, corpus_year, database_type,
     def _corpus_parsing(rawdata_path, parsing_path, database_type):
         if not os.path.exists(parsing_path):
             os.mkdir(parsing_path)
-        parsing_dict, dic_failed = bp.biblio_parser(rawdata_path, database_type)
+        parsing_tup = bp.biblio_parser(rawdata_path, database_type,
+                                       inst_filter_list = None,
+                                       country_affiliations_file_path = institute_affil_file_path,
+                                       inst_types_file_path = inst_types_file_path)
```
Comment (BaptisteRefalo) [inline bmgui/parse_corpus_page.py:300]: Those parameters are not recognized as functions arguments for the biblio_parser() method.

File: bmgui/parse_corpus_page.py [hunk 3]
```diff
@@ -432,7 +439,10 @@ def _deduplicate_corpus_parsing():
                                                      inst_filter_list = institutions_filter_list)
         save_parsing_dict(concat_parsing_dict, concat_parsing_path,
                           item_filename_dict, parsing_save_extent)
-        dedup_parsing_dict  = bp.deduplicate_parsing(concat_parsing_dict)
+        dedup_parsing_dict  = bp.deduplicate_parsing(concat_parsing_dict, 
+                                                     norm_inst_status = False,
+                                                     inst_types_file_path = inst_types_file_path,
+                                                     country_affiliations_file_path = institute_affil_file_path)
```
Comment (BaptisteRefalo) [inline bmgui/parse_corpus_page.py:445]: Those parameters are not recognized as functions arguments for the deduplicate_parsing() method.

**Review by BaptisteRefalo — COMMENTED** [review]
Review done :)
I added some remarks, interrogations and suggestions.

File: bmfuncts/consolidate_pub_list.py [hunk 4]
```diff
@@ -364,12 +364,93 @@ def _build_inst_issn_df(if_db_df, use_col_list):
...
+            for _, inst_row in inst_issn_df.iterrows():
+                inst_journal = inst_row[journal_col].upper()
+                if corpus_journal == inst_journal:
+                    if inst_row[issn_col] != unknown:
+                        corpus_df_bis.loc[corpus_idx, issn_col] = inst_row[issn_col]
+                    elif inst_row[eissn_col] != unknown:
+                        corpus_df_bis.loc[corpus_idx, issn_col] = inst_row[eissn_col]
+                    else:
+                        pass
+    return corpus_df_bis
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:387]: That's fine ! Thanks

File: bmfuncts/pub_analysis.py [hunk 5]
```diff
@@ -1036,8 +1038,10 @@ def _save_formatted_df_to_xlsx(results_path, item_filename, item_df,
             inst_pub_addresses_df = pd.concat([inst_pub_addresses_df, dg])
 
     return_tup = bp.build_norm_raw_institutions(inst_pub_addresses_df,
-                                                inst_types_file_path,
-                                                country_affil_file_path,
+                                                inst_types_file_path = inst_types_file_path,
+                                                country_affiliations_file_path = country_affil_file_path,
+                                                country_towns_file = country_towns_file_alias,
+                                                country_towns_folder_path = institutions_folder_path,
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/pub_analysis.py:1044]: Yes, there have been changes in BiblioParsing lib. So it is mandatory to rebuild your virtual environment using the requirements_dev.txt file attached to the BiblioMeter repository to import the last version of this library. This version of BiblioParsing is still under test in amal_dev branch that will be merged soon with the main branch.
Thanks for rasing this point !

File: bmfuncts/consolidate_pub_list.py [hunk 6]
```diff
@@ -364,12 +364,93 @@ def _build_inst_issn_df(if_db_df, use_col_list):
...
+
+
+def _build_if_dict(if_df, if_year, unknown, issn_col, eissn_col, if_col):
+    issn_if_dict = dict(zip(if_df[if_year][issn_col],
+                            if_df[if_year][if_col]))
+    if unknown in issn_if_dict.keys():
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:393]: That's fine ! Thanks.

File: bmfuncts/consolidate_pub_list.py [hunk 7]
```diff
@@ -551,23 +567,15 @@ def _format_and_save(results_df, df_full_path, add_cols):
...
-    final_order_col_list = [final_year_col, journal_col_alias,
-                            issn_col_alias, eissn_col_alias,
-                            most_recent_year_if_col_name,
-                            year_db_if_col_name,
-                            pub_id_nb_col_alias,
-                            corpus_issn_col_alias,
-                           ]
+    most_recent_year_if_col = current_if_col_alias + ', ' + if_most_recent_year
+    year_db_if_col          = database_if_col_alias + ' ' + corpus_year
+    final_year_col          = year_col_alias[0:5]
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:577]: It is used  in the `_format_missing_df` function and redefined at line 426 but indeed, it should not be defined in `add_if` function (at line 577)where it is not used.
Thanks !

File: bmfuncts/consolidate_pub_list.py [hunk 8]
```diff
@@ -364,12 +364,93 @@ def _build_inst_issn_df(if_db_df, use_col_list):
...
+                    if inst_row[issn_col] != unknown:
+                        corpus_df_bis.loc[corpus_idx, issn_col] = inst_row[issn_col]
+                    elif inst_row[eissn_col] != unknown:
+                        corpus_df_bis.loc[corpus_idx, issn_col] = inst_row[eissn_col]
+                    else:
+                        pass
+    return corpus_df_bis
+
+
+def _build_if_dict(if_df, if_year, unknown, issn_col, eissn_col, if_col):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:390]: To be done asap.

File: bmfuncts/consolidate_pub_list.py [hunk 9]
```diff
@@ -364,12 +364,93 @@ def _build_inst_issn_df(if_db_df, use_col_list):
...
+    if eissn_col in list(if_df[if_year].columns):
+        eissn_if_dict = dict(zip(if_df[if_year][eissn_col],
+                                 if_df[if_year][if_col]))
+        if unknown in eissn_if_dict.keys():
+            del eissn_if_dict[unknown]
+    if_dict = {**issn_if_dict, **eissn_if_dict}
+    return if_dict
+
+
+def _get_id(inst_issn_df, journal_name, journal_col, id_col, unknown):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:405]: To be done asap.

File: bmfuncts/consolidate_pub_list.py [hunk 10]
```diff
@@ -364,12 +364,93 @@ def _build_inst_issn_df(if_db_df, use_col_list):
...
+        id_lower = id_lower_df.to_list()[0]
+    id_upper_df = inst_issn_df[inst_issn_df[journal_col] == journal_name.upper()][id_col]
+    id_upper = unknown
+    if not id_upper_df.empty:
+        id_upper = id_upper_df.to_list()[0]
+    journal_id = list(set([id_lower,id_upper]) - set([unknown]))[0]
+    return journal_id
+
+
+def _format_missing_df(results_df, common_args_tup, add_cols):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/consolidate_pub_list.py:418]: To be done asap.

File: bmfuncts/update_impact_factors.py [hunk 11]
```diff
@@ -35,6 +35,53 @@ def journal_capwords(text):
     return text
 
 
+def _capwords_journal_col(journal_col):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:38]: To be done ASAP.

File: bmfuncts/update_impact_factors.py [hunk 12]
```diff
@@ -35,6 +35,53 @@ def journal_capwords(text):
...
+    return lambda row: journal_capwords(row[journal_col])
+
+
+def _formatting_wb_sheet(institute, org_tup, year, df, wb, first, if_database):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:42]: To be done ASAP.

File: bmfuncts/update_impact_factors.py [hunk 13]
```diff
@@ -35,6 +35,53 @@ def journal_capwords(text):
...
+        wb,ws = mise_en_page(institute, org_tup, df, wb, if_database)
+    else:
+        wb.create_sheet(year)
+        wb.active = wb[year]
+        ws = wb.active
+        wb,_ = mise_en_page(institute, org_tup, df, wb, if_database)
+    return wb
+
+
+def _get_if(missing_file_path, useful_col_list,
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:55]: To be done ASAP.

File: bmfuncts/update_impact_factors.py [hunk 14]
```diff
@@ -35,6 +35,53 @@ def journal_capwords(text):
...
+                missing_df.rename(columns = {col: year_if_col_alias}, inplace = True)
+    missing_df = missing_df[useful_col_list]
+    if not missing_df.empty:
+        missing_df[journal_col] = missing_df.\
+            apply(_capwords_journal_col(journal_col), axis = 1)
+
+    return missing_df
+
+
+def _append_df(df1, df2, col_alias):
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:73]: To be Done ASAP.

File: bmfuncts/update_impact_factors.py [hunk 15]
```diff
@@ -35,6 +35,53 @@ def journal_capwords(text):
     return text
 
 
+def _capwords_journal_col(journal_col):
+    return lambda row: journal_capwords(row[journal_col])
+
+
+def _formatting_wb_sheet(institute, org_tup, year, df, wb, first, if_database):
+    if first:
+        ws = wb.active
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:44]: 'ws' points to the active sheet of the workbook wb. You are right, it is returned in this case by the `mise_en_page` function and line 44 is not useful.
Thanks !

[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:44]: However, we need to set the title of the active sheet ws. I'll move line 45 after calling the `mise_en_page` function. Test of this changes ASAP.

File: bmfuncts/update_impact_factors.py [hunk 16]
```diff
@@ -207,15 +205,33 @@ def _formatting_wb_sheet(year, df, wb, first, if_database):
...
                                   wb, first, if_database)
         first = False
-        dfs_list = [most_recent_year_if_df_to_add,
-                    corpus_year_most_recent_year_if_df_to_add]
-        most_recent_year_if_df_to_add = pd.concat(dfs_list)
-
+    return wb, first, most_recent_year_if_df_to_add
+
+
+def _build_recent_year_if_df(institute,
```
[the remark this answers is not in the document]
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:220]: To be done ASAP.

File: bmfuncts/update_impact_factors.py (hunk 1 above)
Comment (TickyWill) [reply bmfuncts/update_impact_factors.py:270]: To be done ASAP.

File: bmgui/parse_corpus_page.py (hunk 2 above)
Comment (TickyWill) [reply bmgui/parse_corpus_page.py:300]: This is related to the BiblioParsing new version as already mentionned.

File: bmgui/parse_corpus_page.py (hunk 3 above)
Comment (TickyWill) [reply bmgui/parse_corpus_page.py:445]: This is related to the BiblioParsing new version as already mentionned.

**Review by BaptisteRefalo — APPROVED** [review]

**Pull request closed without merging**